File: Graphs/graph-kahn.py
# Topological Sort (Kahn's Algorithm)
# Type: DAG, BFS-based
# Time: O(V + E)
# Space: O(V + E)

from collections import deque, defaultdict

def kahn_topological_sort(graph):
    indegree = defaultdict(int)
    for u in graph:
        for v in graph[u]:
            indegree[v] += 1

    queue = deque([u for u in graph if indegree[u] == 0])
    result = []

    while queue:
        u = queue.popleft()
        result.append(u)
        for v in graph.get(u, []):
            indegree[v] -= 1
            if indegree[v] == 0:
                queue.append(v)

    if len(result) == len(graph):
        return result
    else:
        return "Cycle detected"

File: Graphs/graph-kahn.js
// Topological Sort (Kahn's Algorithm)
// Type: DAG, BFS-based
// Time: O(V + E)
// Space: O(V + E)

function kahnTopologicalSort(graph) {
    const indegree = {};
    const result = [];

    // Initialize indegrees
    for (const node in graph) {
        if (!(node in indegree)) indegree[node] = 0;
        for (const neighbor of graph[node]) {
            indegree[neighbor] = (indegree[neighbor] || 0) + 1;
        }
    }

    const queue = [];
    for (const node in graph) {
        if (indegree[node] === 0) queue.push(node);
    }

    while (queue.length > 0) {
        const u = queue.shift();
        result.push(u);

        for (const v of graph[u]) {
            indegree[v]--;
            if (indegree[v] === 0) queue.push(v);
        }
    }

    if (result.length === Object.keys(graph).length) {
        return result;
    } else {
        return "Cycle detected";
    }
}
